INT-4072 Fix applySequence with state propagation

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4072

When `publishSubscribeChannel` is with `applySequence = true`, a `messageToSend` is overridden with `sequenceDetails` using `MessageBuilder`, therefore a new fresh `Message`.
In case of state propagation, e.g. `SecurityContextPropagationChannelInterceptor`, we just lost the state from the `ThreadStatePropagationChannelInterceptor.MessageWithThreadState` because of new `Message<?>`
- Add into `BroadcastingDispatcher` the logic to delegate `pushSequenceDetails` into `MessageWithThreadState` directly do not lose the `state`
- Make `ThreadStatePropagationChannelInterceptor` as `MessageBuilderFactory`-aware and use it to rebuild an `original` `Message<?>` in the `MessageWithThreadState` to populate `SequenceDetails`

**Cherry-pick to 4.2.x**

artembilan committed:
INT-4072 Fix applySequence with state propagation

JIRA: https://jira.spring.io/browse/INT-4072

When `publishSubscribeChannel` is with `applySequence = true`, a `messageToSend` is overridden with `sequenceDetails` using `MessageBuilder`, therefore a new fresh `Message`.
In case of state propagation, e.g. `SecurityContextPropagationChannelInterceptor`, we just lost the state from the `ThreadStatePropagationChannelInterceptor.MessageWithThreadState` because of new `Message<?>`

* Add into `BroadcastingDispatcher` the logic to delegate `pushSequenceDetails` into `MessageWithThreadState` directly do not lose the `state`
* Make `ThreadStatePropagationChannelInterceptor` as `MessageBuilderFactory`-aware and use it to rebuild an `original` `Message<?>` in the `MessageWithThreadState` to populate `SequenceDetails`

**Cherry-pick to 4.2.x**

artembilan committed:
Provide an explicit order for `publishSubscribeChannel` subscribers

Fixes GH-1847 (https://github.com/spring-projects/spring-integration/issues/1847)

artembilan commented:
The current Travis failure:
```
org.springframework.integration.jdbc.leader.JdbcLockRegistryLeaderInitiatorTests > testDistributedLeaderElection FAILED

    java.lang.AssertionError at JdbcLockRegistryLeaderInitiatorTests.java:138
```
isn't related to the PR changes and indicates that the assert in the target test:
```
assertThat(revoked1.await(10, TimeUnit.SECONDS), is(true));
```
indicates that the embedded H2 has some timing problems in the Travis container.
That's my guessing, though. I don't see any other obvious reason of that waiting failure.

artembilan commented:
This fixes #1847 as well.

artembilan committed:
Fix mutation in the `ThreadStatePropagationChannelInterceptor`

Since `BroadcastingDispatcher` invokes `pushSequenceDetails` for each subscribed handler,
make `MessageWithThreadState` as immutable and return a new instance via `cloneWithSequenceDetails()` method with particular `sequenceDetails`.
Previous mutable solution ended up with the issue of concurrent modification.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/dispatcher/BroadcastingDispatcher.java:
```diff
@@ -158,13 +160,23 @@ public boolean dispatch(Message<?> message) {
-						.fromMessage(message)
-						.pushSequenceDetails(message.getHeaders().getId(), sequenceNumber++, sequenceSize)
-						.build();
+				if (message instanceof ThreadStatePropagationChannelInterceptor.MessageWithThreadState) {
```
Rather than tying the dispatcher to interceptors of this type, I think I'd prefer to see a new interface; something like...

```java
/**
  *  A Message<?> implementation that needs special handling in MBF.fromMessage().
  */
public interface ClonableMessage {

    MessageBuilder cloneMessage(); // createBuilderForClone() ?

}
```

Then, in `MessageBuilder.fromMessage()`...

```java
if (message instanceof ClonableMessage) {
    return message.cloneMessage();
}
else {
    return new MessageBuilder(...);
}
```

Or, we could leave the clone logic in the dispatcher (using the interface), but I think it really belongs in the builder.

artembilan replied:
Ha! I thought here about minimal, quick, unbreakable, backporting fix.
But I see your point.
Seems for me much cleaner solution and enough flexible.
Let me come up with something :smile: 

Thank you!
And welcome back with fresh mind! :wink:

garyrussell replied:
The simple fix is ok for the backport, if you pefer.

artembilan committed:
* Introduce `CloneableMessage` abstraction to let any custom `Message` to return `MessageBuilder` with desired context.
* Introduce `DelegatingMessageBuilder` as an extension of the `MessageBuilder` to let custom `CloneableMessage` to return desired customization.
* Add into `MessageBuilder#fromMessage()` `if` for the `CloneableMessage`
* Add into `MutableMessageBuilder` a `warn` about `CloneableMessage`
* Revert changes in the `BroadcastingDispatcher` in favor of `CloneableMessage` in the `MessageBuilder`
* Redo `ThreadStatePropagationChannelInterceptor#MessageWithThreadState` logic to be based on the `CloneableMessage` and `DelegatingMessageBuilder` extension.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/support/CloneableMessage.java:
```diff
@@ -0,0 +1,33 @@
+
+/**
+ * An aspect to mark a {@link Message} as a clone-aware producer.
+ * The {@link #cloneMessage()} implementation should return a new {@link Message},
```
`s/Message/MessageBuilder/` - will fix on merge if this is the only issue.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/dispatcher/BroadcastingDispatcher.java:
```diff
@@ -158,12 +159,16 @@ public boolean dispatch(Message<?> message) {
 		int sequenceSize = handlers.size();
+		Message<?> messageToSend = message;
+		UUID sequenceId = null;
+		if (this.applySequence) {
```
It's not clear why this is not reverted too.

artembilan replied:
Otherwise we call `message.getHeaders().getId()` for each `MessageHandler` in the loop.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/support/DelegatingMessageBuilder.java:
```diff
@@ -0,0 +1,186 @@
+ * @author Artem Bilan
+ * @since 4.2.9
+ */
+public class DelegatingMessageBuilder<T> extends MessageBuilder<T> {
```
It's not at all clear to me why we need this additional complexity.

artembilan replied:
See `MessageWithThreadState` changes:
```
return new DelegatingMessageBuilder<Object>(messageBuilder) {

		@Override
		public Message<Object> build() {
			Message<Object> newMessage = super.build();
			return new MessageWithThreadState<S>(newMessage, MessageWithThreadState.this.state);
		}

};
```

I wasn't able to find any other way to retain `MessageWithThreadState` instance and provide `MessageBuild` to the caller. :cry:

garyrussell replied:
Never mind - I see - we have to implement `build()`.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/channel/interceptor/ThreadStatePropagationChannelInterceptor.java:
```diff
@@ -113,6 +117,20 @@ public MessageHeaders getHeaders() {
 
 		@Override
+		public MessageBuilder<Object> cloneMessage() {
+			MessageBuilder<Object> messageBuilder = MessageBuilder.fromMessage(this.message);
```
I think we should still use the builder factory here.

artembilan replied:
Eh... The problem with `sequence` and `MutableMessage` that all the subscribers will receive the same message with identical `sequenceNumber`. That's why `MutableMessageBuilder` is fully wrong from `PublishSubscribeChannel`.

That's one.

Here we return `MessageBuilder` anyway. Nobody expect any other abstraction because of `CloneableMessage` contract. So, I'm not sure that it sounds reasonable to "mutate" something here if that doesn't fulfill for contract.

But, if you insist, I can reinstate factory here easily.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/support/MessageBuilder.java:
```diff
@@ -81,9 +81,15 @@ public T getPayload() {
+		}
+		else {
+			return new MessageBuilder<T>(message.getPayload(), message);
+		}
```
Perhaps this logic should be in `DefaultMessageBuilderFactory` instead?

artembilan replied:
I thought about that, but there are many places when `MessageBuilder` is used directly.
And that would be disappointing from end-user perspective to see different behavior from factory and from builder.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/support/MutableMessageBuilder.java:
```diff
@@ -73,6 +78,10 @@ public T getPayload() {
+		if (message instanceof CloneableMessage) {
+			logger.warn("For CloneableMessages consider to use regular MessageBuilder. " +
+					"Otherwise custom Message context may be lost.");
+		}
```
If we move the logic to the builder factories, this would not be needed.

artembilan replied:
See my comment for other builder on the matte.
Plus pay attention that we have here `MutableMessageBuilder` and Java doesn't provide multiple inheritance, therefore for `Mutable` case we should have one more delegating wrapper.

Maybe we drop everything and come back to my original solution? :smile: 

I analyzed `MessageWithThreadState` usage and there is only one case when we need to rebuild it. Only that `applySequence` in the `BroadcastingDispatcher`.

garyrussell replied:
As I said, I am ok with a simple fix for 4.3, but we need a generic solution for 5.0 - but in any case, the dispatcher should not have to be aware of nested inner classes in some subset of channel interceptors.

Hence we need the interface, but I'm ok if you want to put the conditional build logic in the dispatcher - and we can revisit the general builder solution if/when another use case arises.

i.e. the middle ground (for 4.3 and 5.0) is your original solution but with an interface rather than a direct reference to the interceptor inner class.

artembilan replied:
OK. I'll like the current abstract solution better :smile: 

But I want to understand what's wrong with it.
Thanks

garyrussell commented:
I think my biggest concern is that if the original message was Mutable, the new one should be too.

Maybe another approach is to change the interface to `DecoratedMessage`.

Then, in the dispatcher

    messageToSend = // build message as presently using the builder factory
    if (message instance of DecoratedMessage) {
        messageToSend = ((DecoratedMessage)).decorate(message, messageToSend);
    }

artembilan committed:
Introduce `MessageDecorator` contract

Remove `CloneableMessage` aspect and everything around
`MessageWithThreadState` is now `MessageDecorator` and `BroadcastingDispatcher` check if incoming `message` is `MessageDecorator` and performs its `decorateMessage` after `builder`

artembilan commented:
Gary, thank you!
Looks like the `decorator` solution is much cleaner :smile: 

I'm not sure what is wrong with that AMQP test, but that failure:
```
org.springframework.integration.amqp.outbound.AsyncAmqpGatewayTests > testConfirmsAndReturns FAILED

    org.springframework.amqp.AmqpException at AsyncAmqpGatewayTests.java:254
```
fully isn't on assertion:
```
receiver.stop();
```
:cry:

garyrussell commented:
Yes, indeed, much cleaner - merged as 999644a530ae8a2acd0be227c97fc9bd83cacfe9 and cherry-picked.
